pkg/providers/vault/client.go

```diff
@@ -27,32 +29,182 @@ import (
 	"github.com/edgexfoundry/go-mod-secrets/pkg"
 )
 
+const (
+	lookupSelfVaultAPI = "/v1/auth/token/lookup-self"
+	renewSelfVaultAPI  = "/v1/auth/token/renew-self"
+)
+
 // Client defines the behavior for interacting with the Vault REST secret key/value store via HTTP(S).
 type Client struct {
 	HttpConfig SecretConfig
 	HttpCaller Caller
+
+	// internal member variables
+	lc      loggingClient
+	context context.Context
 }
 
-// NewSecretClient constructs a SecretClient which communicates with Vault via HTTP(S)
-func NewSecretClient(config SecretConfig) (pkg.SecretClient, error) {
-	httpClient, err := createHTTPClient(config)
+func (c Client) refreshToken(ctx context.Context, tokenExpiredCallback tokenExpiredCallback) error {
+	tokenData, err := c.getTokenLookupResponseData()
+
 	if err != nil {
-		return Client{}, err
+		return err
 	}
 
-	if config.RetryWaitPeriod != "" {
-		retryTimeDuration, err := time.ParseDuration(config.RetryWaitPeriod)
-		if err != nil {
-			return nil, err
+	if !tokenData.Data.Renewable {
+		// token is not renewable, log warning and return
+		c.lc.Warn("token is not renewable from the secret store")
+		return nil
+	}
+
+	// the renew interval is half of period value
+	tokenPeriod := time.Duration(tokenData.Data.Period) * time.Second
+	renewInterval := tokenPeriod / 2
+	if renewInterval <= 0 {
+		// no renew
+		c.lc.Warn("no token renewal since renewInterval is 0")
+		return nil
+	}
+
+	ttl := time.Duration(tokenData.Data.Ttl) * time.Second
+
+	// if the current time-to-live is already less than the half of period
+	// need to renew the token right away
+	if ttl <= renewInterval {
+		// call renew self api
+		c.lc.Info("ttl already <= half of the renewal period")
+		if err := c.renewToken(); err != nil {
+			return err
 		}
-		config.retryWaitPeriodTime = retryTimeDuration
 	}
 
-	return Client{
-		HttpConfig: config,
-		HttpCaller: httpClient,
-	}, nil
+	c.context = ctx
+
+	// goroutine to periodically renew the service token based on renewInterval
+	go c.doTokenRefreshPeriodically(renewInterval, tokenExpiredCallback)
 
+	return nil
+}
+
+func (c Client) doTokenRefreshPeriodically(renewInterval time.Duration,
+	tokenExpiredCallback tokenExpiredCallback) {
```

Comment:
Can we add a blank line here for readability?

Reply:
Done: thanks for this!